### PR TITLE
Pipeline parallelism point-to-point communication wrapper setup

### PR DESCRIPTION
**Pull request by MicPie**
## Title

- Pipeline parallelism point-to-point communication wrapper setup

## Description

- Draft of the Pipeline parallelism point-to-point communication wrapper setup, that uses only the rank of the current module, the parent module, and the input data to carry out the point-to-point (p2p) communication to support the AWS SageMaker setup.
- The key idea is the following:
AWS SageMaker wants to support arbitrary execution setups, which is why we approach the p2p. com. solution in following way:
1.) We wrap every `nn.Module` object with the `PPModuleWrapper` object by using `wrap_nn_modules`function.
2.) The `PPModuleWrapper` is set up to handle the communication before and after the original `forward` method of every `nn.Module`.
3.) This is done with the `PPPreFwdP2PCom` and `PPPostFwdP2PCom` `torch.autograd.Function`s.
```
   PPPreFwdP2PCom      PPPostFwdP2PCom
-->  pre fwd --> fwd --> post fwd -->
<-- post bwd <-- bwd <--  pre bwd <--
 ```
- What still needs to be done:
1.) Update `module_rank` and `module_rank_parent` argument setup in the p2p com., as it is not currently setup in the `PPModuleWrapper` class.
2.) Setup the `PipelineParallel` according to the `TensorParallel` (see https://github.com/tunib-ai/oslo/blob/75fed710075034be44ed94c055c7882b7334aaae/oslo/torch/nn/parallel/distributed/tensor_parallel/tensor_parallel.py#L11).
3.) Setup test run like in https://github.com/tunib-ai/oslo/blob/75fed710075034be44ed94c055c7882b7334aaae/tests/test_tp_1d.py with `PipelineParallel` and the updated p2 com. setup.
4.) Pipeline schedulers.

## Linked Issues

- resolved #00

### CODE REVIEW

File: oslo/torch/nn/parallel/distributed/pipeline_parallel/p2p.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        self.post_com = PPPostFwdP2PCom(self.rank, self.rank_parent).apply
+        
+    def forward(self, x):
+        x = self.pre_com(x)
+        x = self.module(x)
+        x = self.post_com(x)
+        return x
+
+
+def wrap_nn_modules(m):
```
Comment (hyunwoongko) [inline oslo/torch/nn/parallel/distributed/pipeline_parallel/p2p.py:128]: @MicPie @Quentin-Anthony  Did you test this?

Comment (MicPie) [reply oslo/torch/nn/parallel/distributed/pipeline_parallel/p2p.py:128]: Yes, local test works as shown below:
```
> model = CustomModule(); model
CustomModule(
  (module1): Linear(in_features=8, out_features=8, bias=True)
  (module2): Sequential(
    (0): Linear(in_features=8, out_features=8, bias=True)
    (1): Sequential(
      (0): Linear(in_features=8, out_features=8, bias=True)
      (1): Linear(in_features=8, out_features=8, bias=True)
    )
    (2): Linear(in_features=8, out_features=8, bias=True)
  )
)

> wrap_nn_modules(model); model
CustomModule(
  (module1): PPModuleWrapper(
    (module): Linear(in_features=8, out_features=8, bias=True)
  )
  (module2): PPModuleWrapper(
    (module): Sequential(
      (0): PPModuleWrapper(
        (module): Linear(in_features=8, out_features=8, bias=True)
      )
      (1): PPModuleWrapper(
        (module): Sequential(
          (0): PPModuleWrapper(
            (module): Linear(in_features=8, out_features=8, bias=True)
          )
          (1): PPModuleWrapper(
            (module): Linear(in_features=8, out_features=8, bias=True)
          )
        )
      )
      (2): PPModuleWrapper(
        (module): Linear(in_features=8, out_features=8, bias=True)
      )
    )
  )
)
```

Comment (MicPie) [reply oslo/torch/nn/parallel/distributed/pipeline_parallel/p2p.py:128]: Let me update `oslo/torch/nn/parallel/distributed/pipeline_parallel/test_model.py` to have that included.

Comment (MicPie) [reply oslo/torch/nn/parallel/distributed/pipeline_parallel/p2p.py:128]: Working test pushed.

**Pull request closed without merging**